src/frontend/SignatureMismatch.ml

```diff
@@ -170,34 +189,62 @@ let promote es promotions =
       let open UnsizedType in
       let emeta = exp.emeta in
       match prom with
-      | RealPromotion when is_int_type emeta.type_ ->
+      | IntToRealPromotion when is_int_type emeta.type_ ->
           Ast.
             { expr= Ast.Promotion (exp, UReal, emeta.ad_level)
             ; emeta= {emeta with type_= promote_array emeta.type_ UReal} }
-      | ComplexPromotion when not (is_complex_type emeta.type_) ->
+      | (IntToComplexPromotion | RealToComplexPromotion)
+        when not (is_complex_type emeta.type_) ->
           { expr= Promotion (exp, UComplex, emeta.ad_level)
           ; emeta= {emeta with type_= promote_array emeta.type_ UComplex} }
       | _ -> exp )
 
-let returntype env name args =
+let promotion_cost p =
+  match p with
+  | None -> 0
+  | RealToComplexPromotion | IntToRealPromotion -> 1
+  | IntToComplexPromotion -> 2
+
+let unique_minimum_promotion ps =
+  let size (_, p) =
+    List.fold ~init:0 ~f:(fun acc p -> acc + promotion_cost p) p in
+  let sizes = List.map ~f:size ps in
+  let min_promotions = List.min_elt ~compare:Int.compare sizes in
+  let ps = List.zip_exn sizes ps in
```

Comment:
Can we not reuse the name of the input inside of functions? Just hard to keep track of in my mental model

Reply:
Fair point here, I try not to shadow if the type is also changing